### DISCUSSION TITLE
Specified sleep period exceeds maximum

### DISCUSSION THREAD

**Discussion by hahahahywel** [Chimildic/goofy#248] [Помощь с ошибкой] [answered]
Once again I am getting this error which seemed to resolve itself a few months ago but now it's coming back. A few weeks ago all my triggers failed with the message "Specified sleep period exceeds maximum" and a wait time of about 24000 seconds. I cancelled all of my triggers and didn't do anything for a week. When I returned, I slowly started manually triggering but the error message came back. I then created a new client which seemed to work for a few hours and slowly started adding more of my scripts to the editor but my play history trigger has been failing for the past 12 hours or so. 

Not entirely sure what to do anymore. Any suggestions? Here is the offending script although I think they all fail tbh

```
function PlayHistory() {

// get play history
 let recent_track = RecentTracks.get(30);
 let play_history = Source.getPlaylistTracks('Play History','6J3OdPDClqYrXg7SGlPr1D','',30);



 Filter.dedupTracks(recent_track,5000);

 Filter.removeTracks(recent_track,play_history);



  // add to playlist
  Playlist.saveWithAppend({
    id: '6J3OdPDClqYrXg7SGlPr1D',
    name: 'Play History',
    position: 'begin',
    tracks: recent_track,
    randomCover: 'update',
  });


}
```

**Comment by Chimildic** [comment, marked as answer]
Unfortunately, count limit in `Source.getPlaylistTracks` apply after get all tracks. So if you trigger fires every 15-30 minutes we have many request to Spotify in short period. Therefore API return shadow ban for 24000 seconds in your case. 

As a rule, Spotify has own history page. Possible you dont need to do it via goofy? Or just update such playlist once a day with [saveWithUpdate](https://chimildic.github.io/goofy/#/reference/playlist?id=savewithupdate).

Let me know if you still want to use code from example.

**Reply by hahahahywel** [reply]
Thanks for the suggestion. I've begun updating the playlist less frequently and it's partially working. I don't understand why this is happening now though as everything was working fine a few weeks ago. Perhaps Spotify has changed it's API?

**Comment by Bedmit** [comment]
Тоже зачастила выдача такого бана. 
Несколько раз подряд получал его в runTasks_  Потом все само собой нормализовалось.
![2023-10-18_10-59-52](https://github.com/Chimildic/goofy/assets/76230194/4bb2acb8-f811-4066-90bd-7b0d71f6d821)

Стал сбоить скрипт плейлиста дня. Вчера вышло время, а сегодня сразу завершился с баном
![2023-10-18_11-07-53](https://github.com/Chimildic/goofy/assets/76230194/e15651be-5a41-4651-9c67-6f791bd46cec)

Попытался по отладке посмотреть какая функция провоцирует бан. Оказалось `Source.getArtistsTopTracks`
На отдельном куске кода, как раз на ней уходит в бан
![2023-10-18_11-15-57](https://github.com/Chimildic/goofy/assets/76230194/55297261-1ae9-4755-bb25-db5e25f39712)

При отладке 
![2023-10-18_11-18-39](https://github.com/Chimildic/goofy/assets/76230194/2a4aa41e-3e91-43e0-a0ac-61e5e33091ae)

**Reply by Bedmit** [reply]
В общем, причина получения бана за  функцию `Source.getArtistsTopTracks` выяснилась. 
По логике скрипта, она получала топ артистов лайкнутых в определенный период. Обычно эта цифра исчисляется десятками. А сейчас этим периодом стали те дни, когда я укорачивал, а потом восстанавливал список любимых треков, борясь с ошибкой в `getSavedTracks`. Как раз тогда и получилось, что в короткий период было лайкнуто 3000 треков. Вполне логично, что от такого объема обращений скрипт просто подавился. Тут в принципе вопросов нет. 

Вопрос только почему раньше этих банов не было? Хотя например во времена экспериментов, были довольно объемные и иногда не перевариваемые задачи в скриптах, и ничего. А сейчас Spotify стал за такое отправлять на скамью штрафников. Аж на сутки, как сегодня.  

Решил кстати этот момент переходом на другое приложение в Dashboard. Как раз старая версия и пригодилась. 
Единственное, так пока и не понятны баны в runTasks_

**Comment by StrikS** [comment]
Тоже столкнулся с такой проблемой. У меня она возникает, если второй раз за день использую функцию craftTracks(). На первый раз она срабатывает успешно создавая мне плейлист за 45.736 сек. Вторая попытка была спустя 3 с половиной часа. Бан на день, как я понимаю.
```
let savedTracks = Source.getSavedTracks();
  console.log('Лайки:', savedTracks.length);

  let history = Cache.read('History.json');
  console.log('История:', history.length);

  Order.shuffle(savedTracks);
  let recomTracks = Source.craftTracks(savedTracks, {
    key: 'seed_tracks',
    query: {
        limit: 10, // по умолчанию и максимум 100
        min_popularity: 80,
    }
  });
  console.log('Рекомендованные треки:', recomTracks.length);
```
```
14:54:33 Информация Треки из миксов без дедупликации: 330
14:54:34 Информация Лайки: 601
14:54:35 Информация История: 2322
14:54:35 Информация Операция продолжится после паузы 74039с.
```
<details>
  <summary>API requests by endpoint</summary>

![image](https://github.com/Chimildic/goofy/assets/15943769/d935943e-4709-435e-b2a0-cf1235b985b1)
</details>
